Document new React APIs in 3.9 - Part 1 of 2 (#11512)
apollographql/apollo-client: docs/source/data/suspense.mdx

The `useBackgroundQuery` hook used in a parent component is responsible for kicking off fetches, but doesn't deal with reading or rendering data. This is delegated to the `useReadQuery` hook used in a child component. This separation of concerns provides a nice performance benefit because cache updates are observed by `useReadQuery` and re-render only the child component. You may find this as a useful tool to optimize your component structure to avoid unnecessarily re-rendering parent components when cache data changes.

<MinVersion version="3.9.0">

### Querying in response to user interaction

</MinVersion>

`useSuspenseQuery` and `useBackgroundQuery` are useful hooks that will begin loading data as soon as the hook is mounted. But what happens when you'd like to start loading your query in response to a user interaction? With `useSuspenseQuery` and `useBackgroundQuery`, you'd need to update React state in response to the user interaction in order to re-render your component.

> Starting with Apollo Client `3.9.0`, you can use `useLoadableQuery` to start loading data in response to user interaction. This is not only more ergonomic, but starts loading the query immediately, providing a nice performance benefit.

Reviewer: Personally I would entirely remove it.

To me the focus of this api is more that it triggers the render as part of an interaction or in response to an event, instead of it happening as a result of render - so you don't have to go for the "workaround" with the additional mental load of setting a variable and then having the render do the request.

> Starting with Apollo Client `3.9.0`, you can use `useLoadableQuery` to start loading data in response to user interaction. This is not only more ergonomic, but starts loading the query immediately, providing a nice performance benefit.

Author: Take 2 [4cdcb5b](https://github.com/apollographql/apollo-client/pull/11512/commits/4cdcb5b954c1f1926955a5d46835b8137992b864)

> Starting with Apollo Client `3.9.0`, you can use `useLoadableQuery` to start loading data in response to user interaction. This is not only more ergonomic, but starts loading the query immediately, providing a nice performance benefit.

Reviewer: Love it!

`useLoadableQuery` returns a function that when called will begin fetching the query with the provided variables. Like `useBackgroundQuery`, you get access to a `queryRef` that is passed to `useReadQuery` to read the data in a child component. When the child component renders before the query has finished loading, the child component suspends.

Let's update our example to start loading the dog's details as a result of selecting from a dropdown.

```tsx {3,8,13,22,29}
import {
  // ...
  useLoadableQuery
} from '@apollo/client';

function App() {

>   const { data } = useSuspenseQuery(GET_DOGS_QUERY);
>   const [queryRef, loadDog] = useLoadableQuery(GET_DOG_QUERY)

Reviewer: ```suggestion
  const { data } = useSuspenseQuery(GET_DOGS_QUERY);
  const [queryRef, loadDog] = useLoadableQuery(GET_DOG_QUERY);
```

We need to slowly enable prettier for the docs so we don't have this kind of inconsistency (and don't have to watch out for them)

  return (
    <>
      <select
        onChange={(e) => loadDog({ id: e.target.value })}
      >
        {data.dogs.map(({ id, name }) => (
          <option key={id} value={id}>
            {name}
          </option>
        ))}
      </select>
      <Suspense fallback={<div>Loading...</div>}>
        {queryRef && <Dog queryRef={queryRef} />}
      </Suspense>
    </>
  );
}

function Dog({ queryRef }) {
  const { data } = useReadQuery(queryRef)

  return (
    <>
      <div>Name: {data.dog.name}</div>
      <div>Breed: {data.dog.breed}</div>
    </>
  );
}
```

> It's important to note that the `queryRef` is `null` until the query loading function is called for the first time. You should conditionally render the child component when the query has not yet been loaded.

> We begin fetching our `GET_DOG_QUERY` as soon as a new dog is selected rather than waiting for our `Dog` component to re-render. When the `Dog` component renders, it reads data from the `GET_DOG_QUERY`, and if not ready, will suspend. As a result of this change, we've also removed the need to track the selected dog ID as part of React state!

Reviewer: > than waiting for our `Dog` component to re-render

Again, I'm not sure if this should be the main selling angle here - that render is probably instant, and it will happen either way, so a data fetch would at best happen a handful of milliseconds sooner. I'd really focus more on the data flow.

Author: Take 2 [d7db9a9](https://github.com/apollographql/apollo-client/pull/11512/commits/d7db9a9ab099a141b0686ecdea4d53778eaf1ae1)

The `useBackgroundQuery` hook used in a parent component is responsible for kicking off fetches, but doesn't deal with reading or rendering data. This is delegated to the `useReadQuery` hook used in a child component. This separation of concerns provides a nice performance benefit because cache updates are observed by `useReadQuery` and re-render only the child component. You may find this as a useful tool to optimize your component structure to avoid unnecessarily re-rendering parent components when cache data changes.

<MinVersion version="3.9.0">

### Fetching in response to user interaction

</MinVersion>

> `useSuspenseQuery` and `useBackgroundQuery` are useful hooks that will begin loading data as soon as the hook is mounted. But what happens when you'd like to start loading your query in response to a user interaction? With `useSuspenseQuery` and `useBackgroundQuery`, data fetching is bound to the component render. You would have to trigger a rerender to start fetching data for a different set of variables (e.g. by changing local component state).

Author: I like this a lot more. I think i panicked because this section felt light, but I like how this is worded much more. Appreciate the feedback!

Starting with Apollo Client `3.9.0`, you can use `useLoadableQuery` to start loading data in response to user interaction. This avoids adding additional component state to trigger rerenders and provides a more ergonomic experience.

> `useLoadableQuery` returns a both an execution function and `queryRef`. The execution function begins fetching the query when called with the provided variables. Like `useBackgroundQuery`, passing the `queryRef` to `useReadQuery` in a child component suspends the child component until the query finishes loading.

Author: Yep, I'm happy with the alignment from other sections. Thanks for catching that typo!

<Note>

> The `queryRef` is `null` until the execution function is called for the first time. You should conditionally render the child component when the query has not yet been loaded.

Author: Yep I like that!

</Note>

> Let's update our example to start loading the dog's details as a result of selecting from a dropdown.

Author: Awesome. Thanks so much!

```tsx {3,8,13,22,29}
import {
  // ...
  useLoadableQuery
} from '@apollo/client';

function App() {
  const { data } = useSuspenseQuery(GET_DOGS_QUERY);

>   const [queryRef, loadDog] = useLoadableQuery(GET_DOG_QUERY);

Author: Oh man that would have been confusing to people. Thanks for catching that!

  return (
    <>
      <select
        onChange={(e) => loadDog({ id: e.target.value })}
      >
        {data.dogs.map(({ id, name }) => (
          <option key={id} value={id}>
            {name}
          </option>
        ))}
      </select>
      <Suspense fallback={<div>Loading...</div>}>
        {queryRef && <Dog queryRef={queryRef} />}
      </Suspense>
    </>
  );
}

> function Dog({ queryRef }) {

Author: Ah yes I missed that from previous examples. Good catch!

  const { data } = useReadQuery(queryRef)

  return (
    <>
      <div>Name: {data.dog.name}</div>
      <div>Breed: {data.dog.breed}</div>
    </>
  );
}
```

> We begin fetching our `GET_DOG_QUERY` as soon as a dog is selected by calling the `loadDog` function. When the `Dog` component renders, `useReadQuery` suspends the component while the `GET_DOG_QUERY` loads, then provides `data` to the component. As a result of this change, we've also removed the need to track the selected dog ID as part of React state!

Author: I dig it! I took it a bit further and removed a couple more "we"s in the description as well. Thanks for the suggestion! 

[680714d](https://github.com/apollographql/apollo-client/pull/11512/commits/680714d7aa6105baec30b4ff96a1a353d4f7571e)

<MinVersion version="3.9.0">

### Initiating queries outside React

</MinVersion>

> Starting with Apollo Client `3.9.0`, it is possible to start loading queries outside React. This avoids the need to wait for React to render your components before the query begins loading, providing a nice performance benefit.

Author: Love it

<ExperimentalFeature>

This feature is in [alpha stage](https://www.apollographql.com/docs/resources/product-launch-stages/#alpha--beta) for version `3.9.0` and may be subject to change before `3.10.0`. We consider this feature production-ready, but may be subject to change depending on feedback. If you'd like to provide feedback for this feature before it is stabilized in `3.10.0`, please visit [#11519](https://github.com/apollographql/apollo-client/issues/11519) and add a comment.

</ExperimentalFeature>

> Before you can preload queries, you first need to create a preload function with `createQueryPreloader`. `createQueryPreloader` is given an `ApolloClient` instance as an argument and returns a function that when called, will begin fetching the query. By making `createQueryPreloader` a [curried](https://en.wikipedia.org/wiki/Currying) function, you can easily export your preload function alongside any `ApolloClient` instance you create.

Reviewer 2: ```suggestion
To preload queries, we'll first need to create a `preloadQuery` function using `createQueryPreloader`. `createQueryPreloader` takes an `ApolloClient` instance as an argument, and the function it returns accepts a query document as an argument and initiates a network request when called.

This `preloadQuery` function created by `createQueryPreloader` does not need to be recreated for each preloaded query. In fact, you may prefer to export the function alongside your Apollo Client instance.
```

Experimenting with removing the reference to currying here. I think it can work either way, but I do worry it's additional unnecessary info for a certain user that would detract from the usage pattern we're trying to explain here. Lmk what you think.

Author: I mostly like this 🙂. I agree the part about currying is a bit distracting, so I removed it. Per our writing guidelines, "we" should be avoided when possible, so I'm going to leave this as "you" for now to direct this at the reader. Combining this with existing led me here: [3207653](https://github.com/apollographql/apollo-client/pull/11512/commits/32076535350f6c0f7c582c2e054919579db9624b)

I also took it a step further and added a tip about exporting the function as that was more what I liked about the "currying" aspect of this anyways and should better resonate with the reader. I did a small bit of reorg though, otherwise there were 3 note components all next to each other 😆 

[a135979](https://github.com/apollographql/apollo-client/pull/11512/commits/a135979ae5804a0d6c8dd43ba204a560c3d98561)

Reviewer 2: Nice! Love the final reorg 💯

The preload function returns a `queryRef` that is passed to `useReadQuery` to read the query data and suspend the component while loading. `useReadQuery` will ensure that your component is kept up-to-date with cache updates for the preloaded query.

<Note>

> It is safe to unmount components that contain preloaded queries. Doing so puts the `queryRef` in a dormant state until the component is mounted again. Any cache updates that may have happened while dormant will be read immediately as if the preloaded query had never been unmounted.

Author: I like those edits!

Technically this is true for all hooks that produce `queryRef`s, they are just less noticeable with those hooks (in fact, thats what https://github.com/apollographql/apollo-client/pull/11442 and https://github.com/apollographql/apollo-client/pull/11438 were all about 🙂).

If you're preloading a query outside of React and outside of a router, I wanted to make it clear that you're not just keeping around a reference to something that is going to eat some of your performance by handling cache updates to nothing (i.e. this is what I mean by the "dormant" state. Perhaps I should just call it the technical term "disposed" state? 🤔 ). It feels way different keeping a reference to a `queryRef` outside React, so I just wanted to assure the reader that it is ok to unmount those components.

Does that answer your question?

> It is safe to unmount components that contain preloaded queries. Doing so puts the `queryRef` in a dormant state until the component is mounted again. Any cache updates that may have happened while dormant will be read immediately as if the preloaded query had never been unmounted.

Reviewer 2: > Technically this is true for all hooks that produce queryRefs, they are just less noticeable with those hooks (in fact, thats what https://github.com/apollographql/apollo-client/pull/11442 and https://github.com/apollographql/apollo-client/pull/11438 were all about 🙂).

Ahhh right of course, thanks for that! That's what I thought initially and then I confused myself 😅 I think it was the wording of `It is safe to unmount components that contain preloaded queries.` that made me wonder if there actually were any mechanics specific to preloaded queries, but I may just be too close to it and reading too much into that :)

> It is safe to unmount components that contain preloaded queries. Doing so puts the `queryRef` in a dormant state until the component is mounted again. Any cache updates that may have happened while dormant will be read immediately as if the preloaded query had never been unmounted.

Author: I had one more edit here to make it closer to whats actually happening under the hood. Would you mind looking at this one more time and providing any additional feedback? Your eyes have been super helpful for this so thank you 🙏 

[5966ad2](https://github.com/apollographql/apollo-client/pull/11512/commits/5966ad2df1557ab7ebb81285fbf06a2837bde18c)

> It is safe to unmount components that contain preloaded queries. Doing so puts the `queryRef` in a dormant state until the component is mounted again. Any cache updates that may have happened while dormant will be read immediately as if the preloaded query had never been unmounted.

Reviewer 2: Yes, that LGTM 🚀

export function loader() {
  return preloadQuery(GET_DOGS_QUERY);
}

export function RouteComponent() {
  const queryRef = useLoaderData();
  const { data } = useReadQuery(queryRef);

  return (
    // ...
  );
}
```

> The `loader` function is available in React Router versions 6.4 and above.

React Router calls the `loader` function which we use to begin loading the `GET_DOG_QUERY` query by calling the `preloadQuery` function. The `queryRef` created by `preloadQuery` is returned from the `loader` function making it accessible in the route component. When the route component renders, we access the `queryRef` from the `useLoaderData` hook, which is then passed to `useReadQuery`. We get the benefit of loading our data early in the routing lifecycle, while our route component maintains the ability to rerender with cache updates!

<Note>

The `preloadQuery` function only works with client-side routing. The `queryRef` returned from `preloadQuery` is not serializable across the wire and as such, will not work with routers that fetch on the server such as [Remix](https://remix.run/).

</Note>

#### Preventing route transitions until the query is loaded

> By default, `preloadQuery` works similar to a [deferred loader](https://reactrouter.com/en/main/guides/deferred) in that routes will transition immediately and suspend with `useReadQuery` until the query is loaded. You may instead want to prevent the route from transitioning until the data is fully loaded. Use the `toPromise` method on the `queryRef` to get a promise that resolves when the query is loaded. This promise resolves with the `queryRef` itself, making it easy to use with hooks such as `useLoaderData`.

Author: Ya I like how this reads a lot more. Thank you!

The `useBackgroundQuery` hook used in a parent component is responsible for kicking off fetches, but doesn't deal with reading or rendering data. This is delegated to the `useReadQuery` hook used in a child component. This separation of concerns provides a nice performance benefit because cache updates are observed by `useReadQuery` and re-render only the child component. You may find this as a useful tool to optimize your component structure to avoid unnecessarily re-rendering parent components when cache data changes.

<MinVersion version="3.9.0">

> ### Querying in response to user interaction

Author: Another point of feedback: Does this section include enough information? 

Some other things we could talk about if we want to provide additional info:

* The hook does not take a `variables` option and instead variables are passed to the `loadQuery` function.
* The hook exports a `reset` function that when called, sets the `queryRef` back to `null`

This section is also provided right before pagination/refetching, which uses `useBackgroundQuery` as the example. Should this be moved afterward? Feedback welcome!

The `useBackgroundQuery` hook used in a parent component is responsible for kicking off fetches, but doesn't deal with reading or rendering data. This is delegated to the `useReadQuery` hook used in a child component. This separation of concerns provides a nice performance benefit because cache updates are observed by `useReadQuery` and re-render only the child component. You may find this as a useful tool to optimize your component structure to avoid unnecessarily re-rendering parent components when cache data changes.

<MinVersion version="3.9.0">

### Querying in response to user interaction

</MinVersion>

> `useSuspenseQuery` and `useBackgroundQuery` are useful hooks that will begin loading data as soon as the hook is mounted. But what happens when you'd like to start loading your query in response to a user interaction? With `useSuspenseQuery` and `useBackgroundQuery`, you'd need to update React state in response to the user interaction in order to re-render your component.

Author: I like it. Tweaked your suggestion a little bit in [b3c64bb](https://github.com/apollographql/apollo-client/pull/11512/commits/b3c64bb3efaa8261b868086af516820f587ca157)

Starting with Apollo Client `3.9.0`, you can use `useLoadableQuery` to start loading data in response to user interaction. This is not only more ergonomic, but starts loading the query immediately, providing a nice performance benefit.

> `useLoadableQuery` returns a function that when called will begin fetching the query with the provided variables. Like `useBackgroundQuery`, you get access to a `queryRef` that is passed to `useReadQuery` to read the data in a child component. When the child component renders before the query has finished loading, the child component suspends.

Author: You don't only get back the function, so is there another way to word this that encapsulates this? Is this ok? Suggestions welcome!

Reviewer: In addition to my suggestion, I would consider switching them around, since they really are in the other order.
```suggestion
`useLoadableQuery` returns a `queryRef` (`null` before you triggered a query) that you can pass to `useReadQuery` in a child component like you would do with a `useBackgroundQuery` `queryRef`, and an execution function that when called will begin fetching the query with the provided variables. 
```

Author: I like it. Tweaked your suggestion a little bit in [b3c64bb](https://github.com/apollographql/apollo-client/pull/11512/commits/b3c64bb3efaa8261b868086af516820f587ca157)

Author: Took the suggestion and made a few more tweaks https://github.com/apollographql/apollo-client/pull/11512/commits/958d6bab4d80236d5d1d11acfb5facfa564ac004

Let's update our example to start loading the dog's details as a result of selecting from a dropdown.

```tsx
function App() {

>   const { data } = useSuspenseQuery(GET_DOGS_QUERY);

Author: Note to self: move the `dogs` to be provided via props to avoid the distraction from `useSuspenseQuery` in this example.

You can gain access to `refetch` and `fetchMore` functions by using the `useQueryRefHandlers` hook. This hook integrates with React transitions, giving you the ability to refetch or paginate without showing the loading fallback.

Let's update our example to preload our `GET_BREEDS_QUERY` outside React and utilize `useQueryRefHandlers` to refetch our query.

```tsx {4,7-8,12}
// ...
import {
  // ...
  useQueryRefHandlers,
} from "@apollo/client";

const preloadQuery = createQueryPreloader(client);
const queryRef = preloadQuery(GET_BREEDS_QUERY); 

function App() {
  const [isPending, startTransition] = useTransition();
  const { refetch } = useQueryRefHandlers(queryRef);

  function handleRefetch() {
    startTransition(() => {
      refetch();
    });
  };

  return (
    <Suspense fallback={<div>Loading...</div>}>
      <Dog
        id="3"
        isPending={isPending}
        onRefetch={handleRefetch}
      />
    </Suspense>
  );
}

function Dog({
  id,
  isPending,
  onRefetch,
}: DogProps) {
  const { data } = useSuspenseQuery(GET_DOG_QUERY, {
    variables: { id },
  });

  return (
    <>
      Name: {data.dog.name}
      <Suspense fallback={<div>Loading breeds...</div>}>

>         <Breeds isPending={isPending} />

Reviewer: I don't have any better idea, but having all breeds listed as a child of `Dog` seems pretty confusing to me.

Author: Discussed in person. This is a continuation of the previous example to keep it in the flow of the examples.

Might be good to revisit this example in a future PR to avoid some of the confusion here.

When loading queries [outside React](#initiating-queries-outside-react), the `preloadQuery` function returns a `queryRef` with no access to `refetch` or `fetchMore` functions. This presents a challenge when you need to refetch or paginate the preloaded query.

You can gain access to `refetch` and `fetchMore` functions by using the `useQueryRefHandlers` hook. This hook integrates with React transitions, giving you the ability to refetch or paginate without showing the loading fallback.

Let's update our example to preload our `GET_BREEDS_QUERY` outside React and use the `useQueryRefHandlers` hook to refetch our query.

```tsx {4,7,11}
// ...
import {
  // ...
  useQueryRefHandlers,
} from "@apollo/client";

const queryRef = preloadQuery(GET_BREEDS_QUERY);

function App() {
  const [isPending, startTransition] = useTransition();
  const { refetch } = useQueryRefHandlers(queryRef);

  function handleRefetch() {
    startTransition(() => {
      refetch();
    });
  };

  return (
    <Suspense fallback={<div>Loading...</div>}>
      <Dog
        id="3"
        isPending={isPending}
        onRefetch={handleRefetch}
      />
    </Suspense>
  );
}

// ...
```

> We begin loading `GET_BREEDS_QUERY` outside of React with the `preloadQuery` function. We use the `useQueryRefHandlers` hook to get the `refetch` function which refetches the query when the button is clicked.

Reviewer 2: ```suggestion
We begin loading our `GET_BREEDS_QUERY` outside of React with the `preloadQuery` function. We pass the `queryRef` returned from `preloadQuery` to the `useQueryRefHandlers` hook which provides us with a `refetch` function we can use to refetch the query when the button is clicked.
```

Thoughts on being explicit about the `queryRef` connection here?

Author: Ya I like that!

> #### Accessing query ref handlers produced from other hooks

Reviewer 2: ```suggestion
#### Using `useQueryRefHandlers` with other Suspense hooks
```

Why is this header so hard to word....

Author: I know right?!!?

I like the first half better, but "Suspense hooks" feels a bit too broad since you don't get a `queryRef` from `useSuspenseQuery` and its a suspense hook. Hmmm....

Author: Its still too long, but its a start. [820742a](https://github.com/apollographql/apollo-client/pull/11512/commits/820742a43056e6464cdb56fcb371461085a4b9a8)

will keep noodling on this

Reviewer 2: Oh yeah 😅 I really like that too fwiw 👍 It doesn't appear in the side bar so I think the length is ok

> `useQueryRefHandlers` is not limited to the `preloadQuery` function and can be used with any hook that produces a `queryRef` such as `useBackgroundQuery` or `useLoadableQuery`. This is useful in situations where you need access to the `refetch` and `fetchMore` functions in components where the `queryRef` is passed through deeply.

Reviewer 2: ```suggestion
`useQueryRefHandlers` can also be used in combination with any hook that returns a `queryRef`, such as `useBackgroundQuery` or `useLoadableQuery`. This is useful when you need access to the `refetch` and `fetchMore` functions in components where the `queryRef` is passed through deeply.
```

Author: Way better 🔥

> Let's update our last example back to `useBackgroundQuery` and get the `refetch` function from the `useQueryRefHandlers` hook in the `Dog` component.

Reviewer 2: ```suggestion
Let's update our example to use `useBackgroundQuery` again and see how we can access a `refetch` function in our `Dog` component using `useQueryRefHandlers`:
```

```tsx {15,16,18-22,30}

> function App() {

Reviewer 2: I'm going to throw the rest of these in the codesandbox too 👍

Author: Ah yes thank you! Was on my todo, but hadn't got to it yet!
